File: app/src/main/java/com/cylonid/nativealpha/model/WebApp.java
package com.cylonid.nativealpha.model;


import android.provider.ContactsContract;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.Switch;

import com.cylonid.nativealpha.R;

public class WebApp {

    private String title;
    private String base_url;
    private String last_used_url;
    private Long timestamp_last_used_url;
    private int timeout_last_used_url;
    private int ID;
    private boolean open_url_external;
    private boolean allow_cookies;
    private boolean allow_third_p_cookies;
    private boolean restore_page;
    private boolean allow_js;
    private boolean active_entry;
    private boolean request_desktop;
    private boolean clear_cache;
    private boolean use_adblock;
    private boolean send_savedata_request;
    private boolean block_images;
    private boolean allow_http;
    private String url_on_first_pageload;
    private boolean allow_location_access;

    public WebApp(String url, int id) {
        title = url.replace("http://", "").replace("https://", "").replace("www.", "");
        base_url = url.toLowerCase();
        ID = id;
        open_url_external = true;
        active_entry = true;
        timeout_last_used_url = 10;
        last_used_url = null;
        restore_page = true;
        allow_cookies = true;
        allow_third_p_cookies = false;
        allow_js = true;
        request_desktop = false;
        clear_cache = false;
        use_adblock = false;
        send_savedata_request = false;
        block_images = false;
        allow_http = false;
        url_on_first_pageload = null;
        allow_location_access = false;

    }

    public WebApp(WebApp other) {
        this.title = other.title;
        this.base_url = other.base_url;
        this.last_used_url = other.last_used_url;
        this.timestamp_last_used_url = other.timestamp_last_used_url;
        this.timeout_last_used_url = other.timeout_last_used_url;
        this.ID = other.ID;
        this.open_url_external = other.open_url_external;
        this.allow_cookies = other.allow_cookies;
        this.allow_third_p_cookies = other.allow_third_p_cookies;
        this.restore_page = other.restore_page;
        this.allow_js = other.allow_js;
        this.active_entry = other.active_entry;
        this.request_desktop = other.request_desktop;
        this.clear_cache = other.clear_cache;
        this.use_adblock = other.use_adblock;
        this.send_savedata_request = other.send_savedata_request;
        this.block_images = other.block_images;
        this.allow_http = other.allow_http;
        this.url_on_first_pageload = other.url_on_first_pageload;
        this.allow_location_access = other.allow_location_access;

    }

    public boolean isAllowHttp() {
        return allow_http;
    }

    public void setAllowHttp(boolean allow_http) {
        this.allow_http = allow_http;
    }

    public boolean isSendSavedataRequest() {
        return send_savedata_request;
    }

    public void setSendSavedataRequest(boolean send_savedata_request) {
        this.send_savedata_request = send_savedata_request;
    }

    public boolean isBlockImages() {
        return block_images;
    }

    public void setBlockImages(boolean block_images) {
        this.block_images = block_images;
    }

    public void markInactive() {
        active_entry = false;
        DataManager.getInstance().saveWebAppData();
    }

    public void allowHTTP() {
        allow_http = true;
        DataManager.getInstance().saveWebAppData();
    }

    public void saveCurrentUrl(String url) {
        if (restore_page) {
            last_used_url = url;
            timestamp_last_used_url = System.currentTimeMillis() / 1000;
            DataManager.getInstance().saveWebAppData();
        }
    }

    public String getSingleLineTitle() {
        if (title.length() > 24) {
            String single_line = title.substring(0, 25);
            single_line += " ...";
            return single_line;
        }

        return title;
    }

    public String getLoadableUrl() {

        if (last_used_url == null)
            return base_url;
        else if (restore_page) {

            long current_time_sec = System.currentTimeMillis() / 1000;
            long diff = Math.abs(timestamp_last_used_url - current_time_sec);
            if (diff <= timeout_last_used_url * 60)
                return last_used_url;
        }

        return base_url;
    }

    public int getTimeoutLastUsedUrl() {
        return timeout_last_used_url;
    }

    public void setTimeoutLastUsedUrl(int timeout_last_used_url) {
        this.timeout_last_used_url = timeout_last_used_url;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getBaseUrl() {
        return base_url;
    }

    public void setBaseUrl(String base_url) {
        this.base_url = base_url;
        DataManager.getInstance().saveWebAppData();
    }

    public int getID() {
        return ID;
    }

    public boolean isOpenUrlExternal() {
        return open_url_external;
    }

    public void setOpenUrlExternal(boolean open_url_external) {
        this.open_url_external = open_url_external;
    }

    public boolean isActiveEntry() { return active_entry; }

    public void setActiveEntry(boolean active_entry) {
        this.active_entry = active_entry;
    }

    public boolean isAllowCookies() {
        return allow_cookies;
    }

    public void setAllowCookies(boolean allow_cookies) {
        this.allow_cookies = allow_cookies;
    }

    public boolean isAllowThirdPartyCookies() {
        return allow_third_p_cookies;
    }

    public void setAllowThirdPartyCookies(boolean allow_third_p_cookies) {
        this.allow_third_p_cookies = allow_third_p_cookies;
    }

    public boolean isRestorePage() {
        return restore_page;
    }

    public void setRestorePage(boolean restore_page) {
        this.restore_page = restore_page;
    }

    public boolean isAllowJs() {
        return allow_js;
    }

    public void setAllowJs(boolean allow_js) {
        this.allow_js = allow_js;
    }

    public boolean isRequestDesktop() {
        return request_desktop;
    }

    public void setRequestDesktop(boolean request_desktop) {
        this.request_desktop = request_desktop;
    }

    public boolean isClearCache() {
        return clear_cache;
    }

    public void setClearCache(boolean clear_cache) {
        this.clear_cache = clear_cache;
    }

    public boolean isUseAdblock() {
        return use_adblock;
    }

    public void setUseAdblock(boolean use_adblock) {
        this.use_adblock = use_adblock;
    }

    public boolean isAllowLocationAccess() {
        return allow_location_access;
    }

    public void setAllowLocationAccess(boolean allow_location_access) {
        this.allow_location_access = allow_location_access;
    }

    public void enableLocationAccess() {
        this.allow_location_access = true;
        DataManager.getInstance().saveWebAppData();
    }

    public void setLastUsedUrl(String last_used_url) {
        this.last_used_url = last_used_url;
    }

    public void setTimestampLastUsedUrl(Long timestamp_last_used_url) {
        this.timestamp_last_used_url = timestamp_last_used_url;
    }

    public String getNonNullUrlOnFirstPageload() {
        if (url_on_first_pageload != null)
            return url_on_first_pageload;

        return base_url;
    }

    public String getUrlOnFirstPageload() {
            return url_on_first_pageload;
    }


    public void saveUrlOnFirstPageLoad(String url) {
        url_on_first_pageload = url;
        DataManager.getInstance().saveWebAppData();
    }


    public void onSwitchCookiesChanged(CompoundButton mSwitch, boolean isChecked) {
        Switch third_party_cookies = mSwitch.getRootView().findViewById(R.id.switch3PCookies);
        if (isChecked)
            third_party_cookies.setEnabled(true);
        else {
            third_party_cookies.setEnabled(false);
            third_party_cookies.setChecked(false);
        }
    }
    public void onSwitchJsChanged(CompoundButton mSwitch, boolean isChecked) {
        Switch switchDesktopVersion = mSwitch.getRootView().findViewById(R.id.switchDesktopSite);
        Switch switchAdblock = mSwitch.getRootView().findViewById(R.id.switchAdblock);
        if (isChecked) {
            switchDesktopVersion.setEnabled(true);
            switchAdblock.setEnabled(true);
        } else {
            switchDesktopVersion.setChecked(false);
            switchDesktopVersion.setEnabled(false);
            switchAdblock.setChecked(false);
            switchAdblock.setEnabled(false);
        }
    }

    public void onSwitchRestorePageChanged(CompoundButton mSwitch, boolean isChecked) {
        EditText textTimeout = mSwitch.getRootView().findViewById(R.id.textTimeout);
        if (isChecked) {
            textTimeout.setEnabled(true);
        }
        else
            textTimeout.setEnabled(false);
     }
}

File: app/src/main/kotlin/com/cylonid/nativealpha/model/WebApp.kt
package com.cylonid.nativealpha.model

import android.view.View
import android.widget.*
import com.cylonid.nativealpha.R
import com.cylonid.nativealpha.util.Const
import com.cylonid.nativealpha.util.Utility
import java.util.*

class WebApp {
    val ID: Int
    var baseUrl: String
    var title: String
    var isOverrideGlobalSettings = true
    var isOpenUrlExternal = false
    var isAllowCookies = false
    var isAllowThirdPartyCookies = false
    var isRestorePage = false
    var isAllowJs = false
    var isActiveEntry = false
    var isRequestDesktop = false
    var isClearCache = false
    var isUseAdblock = false
    var isSendSavedataRequest = false
    var isBlockImages = false
    var isAllowHttp = false
    var urlOnFirstPageload: String? = null
    var isAllowLocationAccess = false
    var userAgent: String? = null
    var isUseCustomUserAgent = false
    var isAutoreload = false
    var timeAutoreload = 0
    var isForceDarkMode = false
    var isUseTimespanDarkMode = false
    var timespanDarkModeBegin: String? = "22:00"
    var timespanDarkModeEnd: String? = "06:00"
    var isIgnoreSslErrors = false
    var isShowExpertSettings = false
    var isSafeBrowsing = false
    var isBlockThirdPartyRequests = false
    var containerId: Int = Const.NO_CONTAINER
    var isUseContainer = false
    var isDrmAllowed = false
    var isShowFullscreen = false
    var isKeepAwake = false
    var isCameraPermission = false
    var isMicrophonePermission = false

    constructor(url: String, id: Int) {
        title = url.replace("http://", "").replace("https://", "").replace("www.", "")
        baseUrl = url
        this.ID = id
        initDefaultSettings()
    }

    constructor(other: WebApp) {
        title = other.title
        ID = other.ID
        baseUrl = other.baseUrl
        urlOnFirstPageload = other.urlOnFirstPageload
        isOverrideGlobalSettings = other.isOverrideGlobalSettings
        containerId = other.containerId
        isUseContainer = other.isUseContainer
        copySettings(other)
    }

    //This part of the copy ctor should be callable independently from actual object construction to copy values of the global web app template
    fun copySettings(other: WebApp) {
        isOpenUrlExternal = other.isOpenUrlExternal
        isAllowCookies = other.isAllowCookies
        isAllowThirdPartyCookies = other.isAllowThirdPartyCookies
        isRestorePage = other.isRestorePage
        isAllowJs = other.isAllowJs
        isActiveEntry = other.isActiveEntry
        isRequestDesktop = other.isRequestDesktop
        isClearCache = other.isClearCache
        isUseAdblock = other.isUseAdblock
        isSendSavedataRequest = other.isSendSavedataRequest
        isBlockImages = other.isBlockImages
        isAllowHttp = other.isAllowHttp
        isAllowLocationAccess = other.isAllowLocationAccess
        userAgent = other.userAgent
        isUseCustomUserAgent = other.isUseCustomUserAgent
        isAutoreload = other.isAutoreload
        timeAutoreload = other.timeAutoreload
        isForceDarkMode = other.isForceDarkMode
        isUseTimespanDarkMode = other.isUseTimespanDarkMode
        timespanDarkModeBegin = other.timespanDarkModeBegin
        timespanDarkModeEnd = other.timespanDarkModeEnd
        isIgnoreSslErrors = other.isIgnoreSslErrors
        isShowExpertSettings = other.isShowExpertSettings
        isSafeBrowsing = other.isSafeBrowsing
        isBlockThirdPartyRequests = other.isBlockThirdPartyRequests
        isDrmAllowed = other.isDrmAllowed
        isShowFullscreen = other.isShowFullscreen
        isKeepAwake = other.isKeepAwake
        isCameraPermission = other.isCameraPermission
        isMicrophonePermission = other.isMicrophonePermission
    }

    private fun initDefaultSettings() {
        if (baseUrl.contains("facebook.com")) {
            userAgent = Const.DESKTOP_USER_AGENT
            isUseCustomUserAgent = true
        }
    }

    /*
        This function is used for settings where the ctor needs to have a different setting because
        we want different behaviour for already existing and newly created Web Apps.
            */
    fun applySettingsForNewWebApp() {
        isOverrideGlobalSettings = false
    }

    fun markInactive() {
        isActiveEntry = false
        Utility.deleteShortcuts(Arrays.asList(ID))
    }

    val singleLineTitle: String
        get() {
            if (title.length > 24) {
                var single_line = title.substring(0, 25)
                single_line += " ..."
                return single_line
            }
            return title
        }

    val alphanumericBaseUrl: String
        get() = baseUrl.replace("\\P{Alnum}".toRegex(), "").replace("https", "").replace("http", "").replace("www", "")

    val nonNullUrlOnFirstPageload: String
        get() = if (urlOnFirstPageload != null) urlOnFirstPageload!! else baseUrl

    fun onSwitchCookiesChanged(mSwitch: CompoundButton, isChecked: Boolean) {
        val switchThirdPCookies = mSwitch.rootView.findViewById<Switch>(R.id.switch3PCookies)
        if (isChecked) switchThirdPCookies.isEnabled = true else {
            switchThirdPCookies.isEnabled = false
            switchThirdPCookies.isChecked = false
        }
    }

    fun onSwitchJsChanged(mSwitch: CompoundButton, isChecked: Boolean) {
        val switchDesktopVersion = mSwitch.rootView.findViewById<Switch>(R.id.switchDesktopSite)
        val switchAdblock = mSwitch.rootView.findViewById<Switch>(R.id.switchAdblock)
        if (isChecked) {
            switchDesktopVersion.isEnabled = true
            switchAdblock.isEnabled = true
        } else {
            switchDesktopVersion.isChecked = false
            switchDesktopVersion.isEnabled = false
            switchAdblock.isChecked = false
            switchAdblock.isEnabled = false
        }
    }

    fun onSwitchForceDarkChanged(mSwitch: CompoundButton, isChecked: Boolean) {
        val switchLimit = mSwitch.rootView.findViewById<Switch>(R.id.switchTimeSpanDarkMode)
        val txtBegin = mSwitch.rootView.findViewById<EditText>(R.id.textDarkModeBegin)
        val txtEnd = mSwitch.rootView.findViewById<EditText>(R.id.textDarkModeEnd)
        if (isChecked) {
            switchLimit.isEnabled = true
            txtBegin.isEnabled = true
            txtEnd.isEnabled = true
        } else {
            switchLimit.isChecked = false
            switchLimit.isEnabled = false
            txtBegin.isEnabled = false
            txtEnd.isEnabled = false
        }
    }

    fun onSwitchTimeSpanDarkChanged(mSwitch: CompoundButton, isChecked: Boolean) {
        val lblBegin = mSwitch.rootView.findViewById<TextView>(R.id.lblDarkModeBegin)
        val lblEnd = mSwitch.rootView.findViewById<TextView>(R.id.lblDarkModeEnd)
        val txtBegin = mSwitch.rootView.findViewById<EditText>(R.id.textDarkModeBegin)
        val txtEnd = mSwitch.rootView.findViewById<EditText>(R.id.textDarkModeEnd)
        if (isChecked) {
            lblBegin.isEnabled = true
            lblEnd.isEnabled = true
            txtBegin.isEnabled = true
            txtEnd.isEnabled = true
        } else {
            lblBegin.isEnabled = false
            lblEnd.isEnabled = false
            txtBegin.isEnabled = false
            txtEnd.isEnabled = false
        }
    }

    fun onSwitchUserAgentChanged(mSwitch: CompoundButton, isChecked: Boolean) {
        val txt = mSwitch.rootView.findViewById<EditText>(R.id.textUserAgent)
        val switchDesktopVersion = mSwitch.rootView.findViewById<Switch>(R.id.switchDesktopSite)
        if (isChecked) {
            switchDesktopVersion.isChecked = false
            switchDesktopVersion.isEnabled = false
            txt.isEnabled = true
        } else {
            txt.isEnabled = false
            switchDesktopVersion.isEnabled = true
        }
    }

    fun onSwitchAutoreloadChanged(mSwitch: CompoundButton, isChecked: Boolean) {
        val text = mSwitch.rootView.findViewById<EditText>(R.id.textReloadInterval)
        val label = mSwitch.rootView.findViewById<TextView>(R.id.labelReloadInterval)
        text.isEnabled = isChecked
        label.isEnabled = isChecked
    }

    fun onSwitchExpertSettingsChanged(mSwitch: CompoundButton, isChecked: Boolean) {
        val expertSettings = mSwitch.rootView.findViewById<LinearLayout>(R.id.sectionExpertSettings)
        if (isChecked) expertSettings.visibility = View.VISIBLE else expertSettings.visibility = View.GONE
    }

    fun onSwitchSandboxChanged(mSwitch: CompoundButton?, isChecked: Boolean) {
        containerId = if (isChecked) {
            SandboxManager.getInstance().calculateNextFreeContainerId()
        } else {
            Const.NO_CONTAINER
        }
    }

    fun onSwitchOverrideGlobalSettingsChanged(mSwitch: CompoundButton, isChecked: Boolean) {
        val sectionDetailedWebAppSettings = mSwitch.rootView.findViewById<LinearLayout>(R.id.sectionWebAppDetailSettings)
        Utility.setViewAndChildrenEnabled(sectionDetailedWebAppSettings, isChecked)
    }
}